### PR TITLE
[RFR] Add option to make ReferenceInput choices lazy

### PR DESCRIPTION
**Pull request by ValentinH**
This is actually implementing a similar feature than the one requested in #5898 but for `<ReferenceInput`.

# Why
The goal is to be able to control when the `useGetList` is `enabled` to avoid over-fetching the choices. In our application, we have a user autocomplete component and, as we have hundred of thousands of possible choices, we only want to start showing/fetching the choices if the user inputs at least 2 characters.

# Solution
This PR introduces an `isEnabled` option that is an optional function we use to set the `enabled` option of the `useGetList()`. This function is passed the `filterValues` array.

Example of usage:
```tsx
<ReferenceInput
  label="example"
  reference="users"
  source="user"
  isEnabled={({ q }) => {
    return !!q && q.length > 2
  }}
>
  <AutocompleteInput ... />
</ReferenceInput>
```

# Alternatives
Today, in order to work-around this, we pass a permanent filter `filter={{ is_autocomplete: true }}` that we use on our backend api to return early to avoid hitting the DB. However, we are still making calls for nothing.

<hr/>

I also looked to add a similar prop for the `ReferenceArrayInput` but, as it's using `useGetMatching`, it's a bit more complex.

If you think this option makes sense, I'd be happy to dig more to also add support for it for the `ReferenceArrayInput`.

### CODE REVIEW

File: packages/ra-core/src/controller/input/useReferenceInputController.ts [hunk 1]
```diff
@@ -121,7 +122,10 @@ export const useReferenceInputController = (
         loaded: possibleValuesLoaded,
         loading: possibleValuesLoading,
         error: possibleValuesError,
-    } = useGetList(reference, pagination, sort, filterValues);
+    } = useGetList(reference, pagination, sort, filterValues, {
+        action: 'CUSTOM_QUERY',
+        enabled: isEnabled && isEnabled(filterValues),
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/useReferenceInputController.ts:127]: This won't work when isEnabled isn't set

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Nice! This will be a welcome addition. 

You need to add unit tests and documentation to have this PR ready to merge.

File: packages/ra-core/src/controller/input/useReferenceInputController.ts (hunk 1 above)
Comment (ValentinH) [reply packages/ra-core/src/controller/input/useReferenceInputController.ts:127]: I actually relied on the `enabled` option default value being `true`. However, I can make it `true` explicitly if you think this is clearer.

**Comment by ValentinH** [comment]
OK, happy that you like it. I'll add the tests and docs then.

What do you think about my comments regarding `ReferenceArrayInput` ?

File: docs/Inputs.md [hunk 2]
```diff
@@ -1646,6 +1648,16 @@ The child component receives the following props from `<ReferenceInput>`:
 - `setPagination`: : function to call to update the pagination of the request for possible values
 - `setSort`: function to call to update the sorting of the request for possible values
 
+**Tip** You can make the `getList()` call lazy by using the `isEnabled` prop. This can be useful when using an `AutocompleteInput` on a resource with a lot of data. The following example only starts fetching the options when the query has at least 2 characters:
```
Comment (fzaninotto) [inline docs/Inputs.md:1651]: I think that 'isEnabled' is too vague a name. ReferenceInput does a lot of things, fetching choices is only one of them. I'd prefer something like 'enableGetChoices'

You should also explain what value this prop expects.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Nice! Feel free to do the same for ReferencearrayInput, too.

File: docs/Inputs.md (hunk 2 above)
Comment (ValentinH) [reply docs/Inputs.md]: Done

**Comment by ValentinH** [comment]
From what I see `useReferenceArrayInputController()` uses `useGetMatching()` and the `enabled` option is actually not working properly there: when the `enabled` flag is false, the `loading` prop is stuck to `true` which will show the loading indicator. I need to dig more into this and probably fix the `enabled` prop behaviour for `useGetMatching`.

If you agree, I'll do this in a follow-up PR.

**Comment by ValentinH** [comment]
I just pushed the support for the `enableGetChoices` prop on the `ReferenceArrayInput`.

One small thing that I find misleading is that the `filterValues` always contain the `q` param for `ReferenceInput` but not for `ReferenceArrayInput`. This enforces checking for `q` being present in the  `ReferenceArrayInput` `enableGetChoices` function:

image

Do you think this is something that should be taken care of or is fine like this?

**Comment by ValentinH** [comment]
@fzaninotto do you think this PR is ready to be merged or is it still missing something?

**Comment by ValentinH** [comment]
Why is this being closed?

**Comment by fzaninotto** [comment]
This PR was automatically closed by GitHub, I'm reopening it. Sorry for the noise.

**Comment by ValentinH** [comment]
I think this PR was closed again by Github. I believe it is still valid.

**Comment by fzaninotto** [comment]
You're right, I'm reopening it.

**Comment by dgkm** [comment]
It's a great feature.  Any idea in which release this PR is getting merged and will be available?

**Review by djhi — APPROVED** [review]

**Comment by ValentinH** [comment]
Thanks for merging this PR! 🎉

**Comment by djhi** [comment]
Thanks!

**Pull request merged**